modepy/tools.py

```diff
@@ -415,4 +439,67 @@ def estimate_lebesgue_constant(n, nodes, shape=None, visualize=False):
 # }}}
 
 
+# {{{ tensor product reshaping
+
+@runtime_checkable
+class Reshapeable(Protocol):
+    shape: Tuple[int, ...]
+
+    def reshape(
+            self: "ReshapeableT", *newshape: Tuple[int, ...], order: str
+            ) -> "ReshapeableT":
+        ...
+
+
+ReshapeableT = TypeVar("ReshapeableT", bound=Reshapeable)
+
+
+def reshape_array_for_tensor_product_space(
+        space: TensorProductSpace, ary: ReshapeableT, axis=-1) -> ReshapeableT:
+    """Return a reshaped view of *ary* that exposes the tensor product nature
+    of the space. Axis number *axis* of *ary* must index coefficients
+    corresponding to a tensor-product-structured basis (e.g. modal or nodal
+    coefficients).
+
+    :arg ary: an array with last dimension having a length matching the
+        :attr:`~modepy.FunctionSpace.space_dim` of *space*.
+    :arg result: *ary* reshaped with axis number *axis* replaced by
+        a tuple of dimensions matching the dimensions of the spaces
+        making up the tensor product. Variation of the represented
+        function along a given dimension will be represented by variation
+        of array entries along the corresponding array axis.
+    """
+    if axis < 0:
+        axis += len(ary.shape)
+    if not (0 <= axis < len(ary.shape)):
+        raise ValueError("invalid axis specified")
```

Comment:
Add a check that the last dimension matches `space_dim` before doing the reshape?

Reply:
https://github.com/inducer/modepy/compare/ef309cf2dd6e479823af74348b42144a75d6037a..993ee847d6096bf265d998ab89229bb6927e3a4b